src/App.tsx

```diff
@@ -1,223 +1,271 @@
-    // focus the element with `ref={newTodoField}`
-    if (newTodoField.current) {
-      newTodoField.current.focus();
+  const showError = useCallback((message: string) => {
+    setError({ status: true, message });
+
+    setTimeout(() => {
+      setError({ status: false, message: '' });
+    }, 3000);
+  }, []);
+
+  const loadTodos = useCallback(async () => {
+    if (user) {
+      try {
+        const todosFromServer = await getTodos(user.id);
+
+        setTodos(todosFromServer);
+      } catch {
+        showError('Unable to load todos');
+      }
     }
   }, []);
 
+  const handleAddNewTodo = async (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+
+    if (user) {
+      setIsAdding(true);
+
+      try {
+        if (!todoTitle.trim().length) {
+          showError('Title can\'t be empty');
+          setIsAdding(false);
+
+          return;
+        }
+
+        await createTodo(todoTitle, user.id);
+        await loadTodos();
+
+        setIsAdding(false);
+        setTodoTitle('');
+      } catch {
+        showError('Unable to add a todo');
+        setIsAdding(false);
+      }
+    }
+  };
+
+  const handleDeleteTodo = useCallback(async (todoId: number) => {
+    setDeletedTodoIds(currentTodoIds => [...currentTodoIds, todoId]);
+
+    try {
+      await deleteTodo(todoId);
+      await loadTodos();
+    } catch {
+      showError('Unable to delete a todo');
+      setDeletedTodoIds([]);
+    }
+  }, []);
+
+  const handleDeleteCompletedTodos = useCallback(async () => {
+    try {
+      const completedTodoIds = todos
+        .filter(todo => todo.completed)
+        .map(todo => todo.id);
+
+      setDeletedTodoIds(completedTodoIds);
+
+      await Promise.all(todos.map(async (todo) => {
+        if (todo.completed) {
+          await deleteTodo(todo.id);
+        }
+
+        return todo;
+      }));
+
+      await loadTodos();
+    } catch {
+      showError('Unable to delete a todo');
+      setSelectedTodoIds([]);
+    }
+  }, [todos]);
+
+  const handleToggleTodo = useCallback(
+    async (todoId: number, completed: boolean) => {
+      setSelectedTodoIds(currentTodoIds => [...currentTodoIds, todoId]);
+
+      try {
+        await updateTodo(todoId, { completed: !completed });
+        await loadTodos();
+
+        setSelectedTodoIds(currentTodoIds => (
+          currentTodoIds.filter(id => id !== todoId)
+        ));
+      } catch {
+        showError('Unable to update a todo');
+        setSelectedTodoIds([]);
+      }
+    }, [],
+  );
+
+  const isEachTodoCompleted = useMemo(
+    () => todos.every(todo => todo.completed), [todos],
+  );
+
+  const handleToggleAllTodos = useCallback(async () => {
+    try {
+      await Promise.all(todos.map(async (todo) => {
+        if (!todo.completed || isEachTodoCompleted) {
+          setSelectedTodoIds(currentTodoIds => [...currentTodoIds, todo.id]);
+
+          await updateTodo(todo.id, { completed: !todo.completed });
+        }
+      }));
+
```

Comment:
- `toggleAll` click changes its status to the oppsite one, and set it to all the todos;
A function does not change completed statuses to incomplete ones if the list does not contain all the items in the same status.